src/choleskyQR.jl

```diff
@@ -114,49 +114,94 @@ end
 
 The resulting polynomials are orthonormal on the same domain as `P`. The supplied `P` must be normalized. Accepted inputs for `sqrtw` are the square root of the weight modification as a function or `sqrtW` as an infinite matrix representing multiplication with the function `sqrt(w)` on the basis `P`.
+
+The underlying QR approach allows two methods, one which uses the Q matrix and one which uses the R matrix. To change between methods, an optional argument :Q or :R may be supplied. The default is to use the Q method.
 """
-function qr_jacobimatrix(sqrtw::Function, P)
+function qr_jacobimatrix(sqrtw::Function, P, method = :Q)
     Q = normalized(P)
     x = axes(P,1)
     sqrtW = (Q \ (sqrtw.(x) .* Q))  # Compute weight multiplication via Clenshaw
-    return qr_jacobimatrix(sqrtW, Q)
+    return qr_jacobimatrix(sqrtW, Q, method)
 end
-function qr_jacobimatrix(sqrtW::AbstractMatrix, Q)
+function qr_jacobimatrix(sqrtW::AbstractMatrix, Q, method = :Q)
     isnormalized(Q) || error("Polynomials must be orthonormal")
-    SymTridiagonal(QRJacobiBand{:dv}(sqrtW,Q),QRJacobiBand{:ev}(sqrtW,Q))
+    F = qr(sqrtW)
+    SymTridiagonal(QRJacobiBand{:dv,method}(F,Q),QRJacobiBand{:ev,method}(F,Q))
 end
 
 # The generated Jacobi operators are symmetric tridiagonal, so we store their data in cached bands
-mutable struct QRJacobiBand{dv,T} <: AbstractCachedVector{T}
-    data::Vector{T}       # store band entries, :dv for diagonal, :ev for off-diagonal
-    U::ApplyArray{T}      # store upper triangular conversion matrix (needed to extend available entries)
-    UX::ApplyArray{T}     # store U*X, where X is the Jacobi matrix of the original P (needed to extend available entries)
-    datasize::Int         # size of so-far computed block 
+mutable struct QRJacobiBand{dv,method,T} <: AbstractCachedVector{T}
+    data::Vector{T}             # store band entries, :dv for diagonal, :ev for off-diagonal
+    U                           # store conversion, Q method: stores QR object. R method: only stores R.
+    UX::AbstractMatrix{T}       # Auxilliary matrix. Q method: stores in-progress incomplete modification. R method: stores U*X for efficiency.
+    P::OrthogonalPolynomial{T}  # Remember original polynomials
+    datasize::Int               # size of so-far computed block
 end
 
 # Computes the initial data for the Jacobi operator bands
-function QRJacobiBand{:dv}(sqrtW, P::OrthogonalPolynomial{T}) where T
-    U = qr(sqrtW).R
-    U = ApplyArray(*,Diagonal(sign.(view(U,band(0)))),U)
+function QRJacobiBand{:dv,:Q}(F, P::OrthogonalPolynomial{T}) where T
+    b = 3+bandwidths(F.R)[2]÷2
+    X = jacobimatrix(P)
+        # we fill 2 entries on the first run
+    dv = zeros(T,2)
+        # fill first entry (special case)
+    M = X[1:b,1:b]
+    getindex(F.factors,b,b) # pre-fill cached array
+    v = [one(T);F.factors[1:b,1][2:end]]
+    H = I - F.τ[1]*v*v'
```

Comment:
This is definitely a dense Householder.... might be small dimensional but you are still forming it first and doing `H*M` instead of just doing the cheaper (mostly) in-place:
```julia
M .= M .- F.τ[1] .*  v .* (v'M)
M .= M .- F.τ[1] .*  (M*v) .* v'
```

Reply:
That can't really be the case: micro-allocations like this are extremely expensive. And a fully in-place version should be 2x as fast:
```julia
julia> function inplaceH!(τ, v, M, w)
              mul!(w, M', v)
              M .= M .- τ .* v .* w
              mul!(w, M, v)
              M .= M .- τ .* w .* v'
       end
inplaceH! (generic function with 2 methods)

julia> τ = 4.5; v = randn(2); M = randn(2,2); w = similar(v);

julia> @btime inplaceH!(τ, v, M, w);
  119.387 ns (0 allocations: 0 bytes)

julia> @btime denseH(τ, v, M);
  203.525 ns (4 allocations: 384 bytes)
```

Reply:
Hm alright I'm convinced. I will get the changes done and check it again, maybe my PC was just doing something in the background, the timings are pretty short and thus volatile. Should use BenchmarkTools.